Fix remaining memory leaks in HEMCO

yantosca opened the pull request:
### Name and Institution (Required)

Name: Bob Yantosca
Institution: Harvard + GCST

### Confirm you have reviewed the following documentation

- [x] [Contributing guidelines](https://geos-chem.readthedocs.io/en/stable/help-and-reference/CONTRIBUTING.html)

### Describe the update
**This is a placeholder PR and should not yet be merged.  I have made the base branch `main` but will change it later.**

This is a companion PR to [geoschem/geos-chem#2102](https://github.com/geoschem/geos-chem/issues/2102).  Configuring GEOS-Chem with `-DSANITIZE=y` has revealed the following memory leaks in GEOS-Chem and HEMCO:
```console
=================================================================
==1495473==ERROR: LeakSanitizer: detected memory leaks

Direct leak of 107464 byte(s) in 133 object(s) allocated from:
    #0 0x14815712993f in __interceptor_malloc ../../.././libsanitizer/asan/asan_malloc_linux.cpp:145
    #1 0x1bd566d in __hco_filedata_mod_MOD_filedata_init /n/holyscratch01/jacob_lab/ryantosca/tests/cloudj/test_memory/CodeDir/src/HEMCO/src/Core/hco_filedata_mod.F90:174
    #2 0x202020202020201f  (<unknown module>)

Direct leak of 29248 byte(s) in 8 object(s) allocated from:
    #0 0x14815712993f in __interceptor_malloc ../../.././libsanitizer/asan/asan_malloc_linux.cpp:145
    #1 0x10c438e in __histcontainer_mod_MOD_histcontainer_create /n/holyscratch01/jacob_lab/ryantosca/tests/cloudj/test_memory/CodeDir/src/GEOS-Chem/History/histcontainer_mod.F90:319
    #2 0x202020202020201f  (<unknown module>)

Direct leak of 3720 byte(s) in 1 object(s) allocated from:
    #0 0x14815712993f in __interceptor_malloc ../../.././libsanitizer/asan/asan_malloc_linux.cpp:145
    #1 0x18f6ae1 in __state_chm_mod_MOD_init_state_chm /n/holyscratch01/jacob_lab/ryantosca/tests/cloudj/test_memory/CodeDir/src/GEOS-Chem/Headers/state_chm_mod.F90:803

Direct leak of 3160 byte(s) in 1 object(s) allocated from:
    #0 0x14815712993f in __interceptor_malloc ../../.././libsanitizer/asan/asan_malloc_linux.cpp:145
    #1 0x18fe162 in __state_chm_mod_MOD_init_state_chm /n/holyscratch01/jacob_lab/ryantosca/tests/cloudj/test_memory/CodeDir/src/GEOS-Chem/Headers/state_chm_mod.F90:1027

SUMMARY: AddressSanitizer: 143592 byte(s) leaked in 143 allocation(s).
```

Leak 1 is in HEMCO.  We are still investigating.

The other leaks are being addressed in https://github.com/geoschem/geos-chem/pull/2104.

Tagging @msulprizio @lizziel

### Expected changes
This will be a zero-diff update that will remove memory leaks.  It will not change results.

### Reference(s)
N/A

### Related Github Issue(s)

- See https://github.com/geoschem/geos-chem/pull/2104
- Closes https://github.com/geoschem/geos-chem/issues/2102

yantosca committed:
Return null ptrs from cleanup routines in hco_arr_mod, hco_tidx_mod

src/Core/hco_arr_mod.F90
- Set Arr and ArrVec to null before exiting cleanup routines


src/Core/hco_tidx_mod.F90
- Set individual fields of AlltIDx to NULL after deallocating
- Update comments

Signed-off-by: Bob Yantosca <yantosca@seas.harvard.edu>

yantosca commented:
@msulprizio @lizziel @christophkeller: I now know where the memory leak is being incurred.  In `HEMCO/src/Core/hco_config_mod.F90` there is a FileData object (`Dta`) that is a local variable that gets allocated, and pointed to by the linked list.  

https://github.com/geoschem/HEMCO/blob/ca9b187aabd0b862aeeb094b7b56b214efc966e1/src/Core/hco_config_mod.F90#L945-L1050

The `Dta` object is also needed further down in the loop for those entries in `HEMCO_Config.rc` that use `-` as the file name (i.e. using the file that is immediately above it).

I tried to replace `Dta` with `Lct%Dct%Dta`, which would get rid of the memory leak, but then the `Dta` object would not be present for those entries that use `-`.

I just wanted to note this.  It might be going deep into the rabbit hole to try to undo this memory leak, so I will probably put this off for the time being.  If any of you have suggestions then let me know.  Thanks!

lizziel commented:
@yantosca, are there memory leak fixes in this PR that are not yet merged? If yes, I wonder if we should merge them now. If there are remaining leaks we can open new PRs for them later.

yantosca commented:
@lizziel, yes maybe we should merge what is present.  I forget if these fix any leaks or if they are just preventative fixes.  I'll rebase atop the current HEMCO.

yantosca commented:
@lizziel @msulprizio: I rebased this branch atop the dev/no-diff-to-benchmark.  I confirmed that this does not introduce any differences in a short simulation. 

The only diffrerences now are that we have added some `==> NULL()` statements to free pointers.  I think we can merge this into 14.8.0 / HEMCO 3.13.0.  Feel free to review at your convenience.

msulprizio reviewed (approved):
Looks good. I noticed there isn't a comment in CHANGELOG.md so you may want to add that.

yantosca commented:
Thanks @msulprizio!  I'll merge into the no-diff-to-benchmark hemco branch

yantosca commented:
All GEOS-Chem Classic integration tests passed:
```console
==============================================================================
GEOS-Chem Classic: Execution Test Results

CodeDir   : 37ab93b GEOS-Chem update: Merge PR #3305 (Fix subsetting bug in GCC History)
GEOS-Chem : 843de4e7d Merge PR #3305 (Fix bug in subsetting of GC-Classic History collections)
HEMCO     : feb3894 Merge PR #255 (Fix remaining memory leaks in HEMCO)
Cloud-J   : 16d18b0 Merge PR #34 (Fixed issues in format statements for verbose output)
HETP      : 2a99b24 Merge pull request #2 from geoschem/bugfix/initialize_local_variables

Using 24 OpenMP threads
Number of execution tests: 38

Submitted as SLURM job: 13654898
==============================================================================


%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%%  All execution tests passed!  %%%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
```
All GEOS-Chem Classic tests were zero-diff w/r/t `gcc.0.7b7ab8e`, except

- gc_4x5_merra2_fullchem_APM
  - Parallel error (known issue)
- The fix for CH4_BBN in PR #3283 causes diffs in the `geoschem_species_metadata.yml` files in these tests.  But diagnostic and restart files are not affected.
  - gc_05x0625_NA_47L_merra2_fullchem
  - gc_2x25_ModelE2.1_fullchem
  - gc_4x5_47L_merra2_fullchem
  - gc_4x5_47L_merra2_fullchem_TOMAS15
  - gc_4x5_47L_merra2_fullchem_TOMAS15_dryrun
  - gc_4x5_merra2_carbon
  - gc_4x5_merra2_carbon_CH4
  - gc_4x5_merra2_carbon_CH4_straddle_00z
  - gc_4x5_merra2_carbon_CO
  - gc_4x5_merra2_carbon_CO2
  - gc_4x5_merra2_carbon_dryrun
  - gc_4x5_merra2_carbon_OCS
  - gc_4x5_merra2_fullchem
  - gc_4x5_merra2_fullchem_aciduptake
  - gc_4x5_merra2_fullchem_alldiags
  - gc_4x5_merra2_fullchem_APM
  - gc_4x5_merra2_fullchem_APM_dryrun
  - gc_4x5_merra2_fullchem_benchmark
  - gc_4x5_merra2_fullchem_benchmark_dryrun
  - gc_4x5_merra2_fullchem_complexSOA
  - gc_4x5_merra2_fullchem_complexSOA_SVPOA
  - gc_4x5_merra2_fullchem_dryrun
  - gc_4x5_merra2_fullchem_LuoWd
  - gc_4x5_merra2_fullchem_marinePOA
  - gc_4x5_merra2_fullchem_RRTMG
  - gc_4x5_merra2_fullchem_RRTMG_dryrun
  - gc_4x5_merra2_Hg
  - gc_4x5_merra2_Hg_dryrun

All GCHP integration tests passed:
```console
==============================================================================
GCHP: Execution Test Results

CodeDir       : bc261ef GEOS-Chem update: Merge PR #3305 (Fix subsetting bug in GCC History)
MAPL          : 80d66d0a Merge PR #41 (ExtData2G in GCHP Transport Tracers simulation)
GMAO_Shared   : 4ddb3ec Merge pull request #2 from geoschem/feature/mapl-upgrade
ESMA_cmake    : 01bbf2f Merge pull request #9 from geoschem/feature/turn_off_f2py_builds_in_gchp
FMS           : 259759d1 Merge pull request #3 from geoschem/feature/update_gmao_libs
FVdycoreCubed : 9d86304 Merge PR #13 (Add option to get upwards mass flux as real4, not just real8)
geos-chem     : 843de4e7d Merge PR #3305 (Fix bug in subsetting of GC-Classic History collections)
HEMCO         : feb3894 Merge PR #255 (Fix remaining memory leaks in HEMCO)
Cloud-J       : 16d18b0 Merge PR #34 (Fixed issues in format statements for verbose output)
HETP          : 2a99b24 Merge pull request #2 from geoschem/bugfix/initialize_local_variables
GFE           : bac529f Merge pull request #40 from Goddard-Fortran-Ecosystem/feature/mathomp4/update-gftl-1.15.2

Number of execution tests: 12

Submitted as SLURM job: 13650129
==============================================================================

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%%  All execution tests passed!  %%%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
```
All GCHP tests were zero-diff w/r/t `gchp.0.9e64be6`

- The fix for CH4_BBN in PR #3283 causes diffs in the `geoschem_species_metadata.yml` files in these tests.  But diagnostic and restart files are not affected.
  - gchp_merra2_carbon
  - gchp_merra2_carbon_CH4
  - gchp_merra2_carbon_CO
  - gchp_merra2_carbon_CO2
  - gchp_merra2_carbon_OCS
  - gchp_merra2_fullchem
  - gchp_merra2_fullchem_alldiags
  - gchp_merra2_fullchem_benchmark
  - gchp_merra2_fullchem_RRTMG
  - gchp_merra2_fullchem_TOMAS15
